server/src/main/java/com/linecorp/centraldogma/server/internal/api/auth/AbstractPermissionCheckingDecorator.java

```diff
@@ -0,0 +1,131 @@
+abstract class AbstractPermissionCheckingDecorator
+        implements DecoratingServiceFunction<HttpRequest, HttpResponse> {
+
+    private static final Set<Permission> ADMIN_PERMISSION = EnumSet.allOf(Permission.class);
+
+    @Override
+    public final HttpResponse serve(Service<HttpRequest, HttpResponse> delegate,
+                                    ServiceRequestContext ctx,
+                                    HttpRequest req) throws Exception {
+
+        final MetadataService mds = MetadataServiceInjector.getMetadataService(ctx);
+        final User user = AuthenticationUtil.currentUser(ctx);
+
+        final String projectName = ctx.pathParam("projectName");
+        checkArgument(!isNullOrEmpty(projectName), "no project name is specified");
+        final String repoName = ctx.pathParam("repoName");
+        checkArgument(!isNullOrEmpty(repoName), "no repository name is specified");
+
+        if (Project.REPO_META.equals(repoName)) {
+            return serveMetaRepo(delegate, ctx, req, mds, user, projectName);
+        } else {
+            return serveNonMetaRepo(delegate, ctx, req, mds, user, projectName, repoName);
+        }
+    }
+
+    private static HttpResponse serveMetaRepo(Service<HttpRequest, HttpResponse> delegate,
+                                              ServiceRequestContext ctx, HttpRequest req,
+                                              MetadataService mds, User user,
+                                              String projectName) throws Exception {
+        if (user.isAdmin()) {
+            return delegate.serve(ctx, req);
+        }
+        // We do not manage permission for "meta" repository. Actually we do not have a metadata of that.
+        // So we need to check "role" here when the request is accessing to "meta" repository.
+        return HttpResponse.from(mds.findRole(projectName, user).handle((role, cause) -> {
+            if (cause != null) {
+                return handleException(cause);
+            }
+            if (role != ProjectRole.OWNER) {
+                throw HttpStatusException.of(HttpStatus.FORBIDDEN);
+            }
+            try {
+                return delegate.serve(ctx, req);
+            } catch (Exception e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        }));
+    }
+
+    private HttpResponse serveNonMetaRepo(Service<HttpRequest, HttpResponse> delegate,
+                                          ServiceRequestContext ctx, HttpRequest req,
+                                          MetadataService mds, User user,
+                                          String projectName, String repoName) throws Exception {
+        final CompletionStage<Collection<Permission>> f;
+        try {
+            if (user.isAdmin()) {
+                f = CompletableFuture.completedFuture(ADMIN_PERMISSION);
+            } else if (user instanceof UserWithToken) {
+                f = mds.findPermission(projectName, repoName, (UserWithToken) user);
```

Comment:
How about moving `user instanceof UserWithToken` to `MetadataService.findPermissions()` so that a caller doesn't need to downcast explicitly?